plugin-server/src/utils/utils.ts

```diff
@@ -678,3 +678,29 @@ export const areMapsEqual = <K, V>(map1: Map<K, V>, map2: Map<K, V>): boolean =>
     }
     return true
 }
+
+/**
+ * Recursively freeze an object and all its properties.
+ *
+ * The most common use case for this is to test that a function does not mutate its arguments. A frozen
+ * object will throw an error if any of its properties are modified.
+ */
+export function deepFreeze<T extends object>(t: T): T {
+    function deepFreezeInner(obj: any) {
+        if (obj === null || typeof obj !== 'object') {
+            return
+        }
+        if (Object.isFrozen(obj)) {
+            return
+        }
+        Object.freeze(obj)
+        Object.keys(obj).forEach((key) => {
+            if (key in obj && obj[key] !== null && typeof obj[key] === 'object' && !Object.isFrozen(obj[key])) {
+                deepFreezeInner(obj[key])
+            }
```

Comment:
I know this is just used in tests, but we're doing double work by checking the type and frozenness of this entry both here and then inside the recursive call

```suggestion
            if (key in obj) {
                deepFreezeInner(obj[key])
            }
```

Reply:
Fixed!